### DISCUSSION TITLE
Pulling TMB data from cBioPortal

### DISCUSSION THREAD

**Discussion by IJaljuli** [karissawhiting/cbioportalR#54] [Q&A]
Is there a way to pull all data sets with TMB to R in uniformly format, just like when downloading multiple selected data sets from cBioPortal.org? 
I downloaded all datasets with variables that has the acronyms TMB using cbioportalR::get_clinical_by_study(). The problem is sometimes columns with similar content have different names. I was wondering if there's a command to download unified (combined) study data to R, where variables with similar content (but different names) are automatically unified.
The same can apply to variables that indicate if the patient was treated (namely immunotherapy ) as well. 
much appreciated!
Iman

**Comment by karissawhiting** [comment]
Hi @IJaljuli!

Thanks for posting this question.  I've included some code below for pulling and unifying TMB data _(although please note that some TMB data may not be directly comparable across studies as it depends on how TMB was defined for the project and what gene panels were used)._ 

As for other types of data like treatment data, although cBioPortal tries to set up standards for how data is stored and represented, I think each study is fairly free to define their own data schema, especially for clinical data. You can use the code I included below to try to find other columns across studies, for example: `str_detect(clinicalAttributeId, "TREATMENT")` but there isn't a way (that I'm aware of) to pull and automatically unify this data across all studies, especially because these fields may be defined differently depending on the study (for example one study may include age at diagnosis as a variable but another may use age at resection, etc).

Hope this helps! Please let me know if I'm not understanding your question correctly.

**Code Example**:

``` r
library(cbioportalR)
library(tidyverse)
library(cbioportalR)

set_cbioportal_db("public")
#> ✔ You are successfully connected!
#> ✔ base_url for this R session is now set to "www.cbioportal.org/api"

# get names of all studies
all_studies <- available_studies()

# get all clinical attributes available for each study
# takes ~1 min
all_clinical_attributes <- map_df(all_studies$studyId, ~ available_clinical_attributes(.x))

# filter TMB attributes only to get correct name of TMB element
tmb_clinical_attributes <- all_clinical_attributes %>%
  filter(str_detect(clinicalAttributeId, "TMB|tmb")) %>%
  select(studyId, clinicalAttributeId) %>%
  distinct()

# This grabs the TMB data for the first 5 studies only as an example (may take a while to pull all)
tmb_data <- pmap_df(
  tmb_clinical_attributes[1:5, ],
  ~ get_clinical_by_study(study_id = .x, clinical_attribute = .y)
)

str(tmb_data)
#> tibble [692 × 8] (S3: tbl_df/tbl/data.frame)
#>  $ uniqueSampleKey    : chr [1:692] "VENHQS1PUi1BNUoxLTAxOmFjY190Y2dh" "VENHQS1PUi1BNUoyLTAxOmFjY190Y2dh" "VENHQS1PUi1BNUozLTAxOmFjY190Y2dh" "VENHQS1PUi1BNUo0LTAxOmFjY190Y2dh" ...
#>  $ uniquePatientKey   : chr [1:692] "VENHQS1PUi1BNUoxOmFjY190Y2dh" "VENHQS1PUi1BNUoyOmFjY190Y2dh" "VENHQS1PUi1BNUozOmFjY190Y2dh" "VENHQS1PUi1BNUo0OmFjY190Y2dh" ...
#>  $ sampleId           : chr [1:692] "TCGA-OR-A5J1-01" "TCGA-OR-A5J2-01" "TCGA-OR-A5J3-01" "TCGA-OR-A5J4-01" ...
#>  $ patientId          : chr [1:692] "TCGA-OR-A5J1" "TCGA-OR-A5J2" "TCGA-OR-A5J3" "TCGA-OR-A5J4" ...
#>  $ studyId            : chr [1:692] "acc_tcga" "acc_tcga" "acc_tcga" "acc_tcga" ...
#>  $ clinicalAttributeId: chr [1:692] "TMB_NONSYNONYMOUS" "TMB_NONSYNONYMOUS" "TMB_NONSYNONYMOUS" "TMB_NONSYNONYMOUS" ...
#>  $ value              : chr [1:692] "1.26666666667" "2.73333333333" "2.33333333333" "5.66666666667" ...
#>  $ dataLevel          : chr [1:692] "SAMPLE" "SAMPLE" "SAMPLE" "SAMPLE" ...
```

<sup>Created on 2022-10-10 with [reprex v2.0.2](https://reprex.tidyverse.org)</sup>

**Reply by IJaljuli** [reply]
Thanks Karissa for the elegant code!
I can see that the dataset downloaded has several variables enclosing "TMB", e.g. 
` unique(tmb_clinical_attributes$clinicalAttributeId)
[1] "TMB_NONSYNONYMOUS"            "TMB"                          "CFDNA_TMB"                   
[4] "TUMOR_TMB"                    "CVR_TMB_COHORT_PERCENTILE"    "CVR_TMB_SCORE"               
[7] "CVR_TMB_TT_COHORT_PERCENTILE" "IMPACT_TMB_SCORE" `

I assume that  "TMB_NONSYNONYMOUS", "TMB" and  "IMPACT_TMB_SCORE" are all the same variable: tumor mutation number ( which I can only verify that only by reading the clinical attribute description, best case scenario).  I can merge them all by recoding them to a uniform name, but I was hoping that there is an automated way to do so. For example, possibly when researchers uploaded the data they gave keys to indicate a uniform name for the clinical variable. 

Hope I was clear enough,
Thanks!
Iman

**Reply by karissawhiting** [reply]
@IJaljuli yea unfortunately I don't think there is a uniform key between data sets for these types of data points.

IMPACT_TMB_SCORE may be normalized in some way? Additionally, some of the TMB measures were likely calculated on larger gene panels, and some on IMPACT gene panels which can result in inherent differences in the calculations as the targeted panels are more sensitive. You can look at the GENE_PANEL attribute for your samples to see which gene panels TMB is calculated on.

Hope this is helpful! Feel free to reach out with any additional questions.

**Reply by IJaljuli** [reply]
This was in fact really helpful, thanks @karissawhiting!
